build-tools/packages/build-tools/src/typeValidator/packageJson.ts

```diff
@@ -88,50 +109,52 @@ export async function getAndUpdatePackageDetails(
         return { skipReason: "Skipping package: type validation disabled" };
     }
 
-    // normalize the version to remove any pre-release version info,
-    // as we shouldn't change the type validation version for pre-release versions
-    const normalizedVersion = packageDetails.pkg.version.includes("-")
-        ? packageDetails.pkg.version.substring(0, packageDetails.pkg.version.indexOf("-"))
-        : packageDetails.pkg.version;
-
-    if (
-        updateOptions === undefined ||
-        normalizedVersion === packageDetails.pkg.typeValidation?.version
-    ) {
-        return packageDetails;
-    }
+    const version = packageDetails.pkg.version;
+    const scheme = detectVersionScheme(version);
+    let previousVersion: ReleaseVersion;
 
-    /*
-     * this is where we build the previous version we use to compare against the current version.
-     * For major we use semver such that the current major is compared against the latest
-     * previous major. In the case of minor we target specific previous minor version.
-     * Similarly for patch we do not use semver, we compare directly to the previous patch version.
-     *
-     * We do this to align with our release process. We are strictest between patches and minor, and looser between major
-     * We may need to adjust this as we adjust our release processes.
-     */
-    let normalizeParts = normalizedVersion.split(".").map((p) => Number.parseInt(p));
-    const validationVersionParts = packageDetails.pkg.typeValidation?.version
-        ?.split(".")
-        .map((p) => Number.parseInt(p)) ?? [0, 0, 0];
-    let semVer = "";
-    if (normalizeParts[0] !== validationVersionParts[0]) {
-        semVer = "^";
-        normalizeParts = [normalizeParts[0] - 1, 0, 0];
-    } else if (normalizeParts[1] !== validationVersionParts[1]) {
-        normalizeParts = [normalizeParts[0], normalizeParts[1] - 1, 0];
+    if (scheme === "internal") {
+        const [pubVer, intVer] = fromInternalScheme(version);
+        if (intVer.major === 0) {
+            throw new Error(`Internal major unexpectedly 0.`);
+        }
+
+        if (intVer.major === 1) {
+            previousVersion = "1.0.0";
+        } else {
+            previousVersion = toInternalScheme(pubVer, `${intVer.major - 1}.0.0`).version;
+        }
+    } else if (scheme === "virtualPatch") {
+        const ver = fromVirtualPatchScheme(version);
+        if (ver.major <= 1) {
+            throw new Error(`Virtual patch major unexpectedly <= 1.`);
+        }
+        previousVersion = toVirtualPatchScheme(`${ver.major - 1}.0.0`).version;
     } else {
-        normalizeParts[2] = validationVersionParts[2];
+        const ver = semver.parse(version);
+        if (ver === null) {
+            throw new Error(`COuldn't parse version string: ${version}`);
+        }
+
+        if (ver.major <= 1) {
+            throw new Error(`Virtual patch major unexpectedly <= 1.`);
+        }
+
+        previousVersion = `${ver.major}.0.0`;
```

Comment:
Similar here, it seems like you are targeting specific version always? How does the branch-specific "previous" calculation play out?

Reply:
Yes. This change makes it deterministic. It will always be the latest major.

> This change updates the typetest code to set the previous version to the exact latest major release, accounting for the Fluid internal version scheme.

Reply:
There is no longer branch specific behavior, though that was always broken in the existing code.

We make promises based on our major releases. Testing against minor versions makes no sense for type compat because we don't make breaking changes in minors.

Just talking client for a sec, IMO, we should always use the latest major and pin it. ZERO changes to the type test config or lock files should be needed before or after a minor or patch releases. Thus zero changes to type tests should be made outside next, where the breaking change happens.

There are still cases where a change might be made in a minor that will change type test config, like adding a new optional interface property or something. With a pinned version, as time goes on, the type test config will become a record of changes between the pinned version and the current version.

I don't follow the reasoning for doing minor versions, but even if we do this, we should still pin versions IMO. At the very least it works around the lockfile/npm problems we have atm.

For individual packages the story is basically the same, but there are some unique issues there that I put in a deck last night. I'll send you a link separately.

Reply:
given issues we were seeing with lockfile/npm, yes agree on pinning the versions. Our lockfiles are not helping us here.

Ideally, yes. Breaking changes are across majors only. But given our typetests may not be providing us full type safety, as well the fact that our bug fixes may be breaking at times (for example, we fixed interface definitions that were not correctly communicating events being emitted), it is much easier to reason about narrower incremental diff/change when you run into typetests failures, then look for compatibility/diff across branches (major releases). Either way, similar logic that applies for patches (release branches) would apply for minor changes (main)?

We could chat about this offline. I don't have high confidence in our typetests right now, so some of these incremental improvements (pinning the version) are definitely welcome.